design/metadata-handling.md

```diff
@@ -0,0 +1,456 @@
+  The index starts from 0.
+- **indexWithOffset** : takes an integer as parameter and returns the sum of
+  the index and the offset parameter
+- **indexWithStep** : takes an integer as parameter and returns the
+  multiplication of the index and the step parameter
+- **indexWithOffsetAndStep** OR **indexWithStepAndOffset**: takes two
+  integers as parameters, order depending on the function name, and returns the
+  sum of the offset and the multiplication of the index and the step.
+- **index*Hex** : All the `index` functions can be suffixed with `Hex` to
+  get the same value in hexadecimal format.
+- **bareMetalHostMACByName**: takes a string as parameter and returns the MAC
+  address of the nic with the name matching the parameter. This function
+  operates over the list of NICs in the `status.hardwareDetails.NIC` field of
+  the BareMetalHost.
+
+
+The output of the controller would be a Metal3Data object,one per node linking to the
+Metal3DataTemplate object and the associated secrets
+
+The Metal3Data object would be:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Data
+metadata:
+  name: nodepool-1-0
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3DataTemplate
+    name: nodepool-1
+spec:
+  index: 0
+  metaData:
+    name: machine-1-metadata
+    namespace: default
+  networkData:
+    name: machine-1-metadata
+    namespace: default
+  metal3Machine:
+    name: machine-1
+    namespace: default
+```
+The Metal3Data will contain the index of this node, and links to the secrets
+generated and to the Metal3Machine using this Metal3Data object.
+
+If the Metal3DataTemplate object is updated, the generated secrets will not be
+updated, to allow for reprovisioning of the nodes in the exact same state as
+they were initially provisioned. Hence, to do an update, it is necessary to do
+a rolling upgrade of all nodes.
+
+The reconciliation of the Metal3DataTemplate object will also be triggered by
+changes on Metal3Machines. In the case that a Metal3Machine gets modified, if
+the `dataTemplate` references a Metal3DataTemplate, that object will be reconciled.
+There will be two cases:
+
+- An already generated Metal3Data object exists with an ownerReference to this
+  Metal3Machine. In that case, the reconciler will verify that the required
+  secrets exist. If they do not, they will be created.
+- if no secret exists with an ownerReference to this Metal3Machine, then the
+  reconciler will create one and fill the respective field with the secret name.
+
+To create a Metal3Data object, the controller will select an index for that
+Metal3Machine. The selection happens by selecting the lowest available index.
+The controller will list all existing Metal3Data object linked to this
+Metal3DataTemplate and create a list of unavailable indexes. It will fill it by
+extracting the index from the Metal3Data names.
```

Comment:
What happens if a hole is created in the list? Say I create 3 hosts, then delete the second. Is index `1` available after that? Or is the next lowest index `3`?

Reply:
yes, it will select 1 as this is the lowest index.

About the additional DSL, I don't really see yet how we could do that with the same level of flexibility. I can work on defining more use-cases so that we can discuss it further

Reply:
OK, it would be good to be explicit about that case. The algorithm is to pick the lowest unallocated number starting with the offset and incrementing by the step, right?

Reply:
the index itself, does not have an offset or a step. it will only increment by 1. The rendering of the index is what takes an offset and a step. I'll make it clearer